analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarCompilationStartAnalysisContext.cs

```diff
@@ -42,22 +43,19 @@ public void RegisterNodeAction<TSyntaxKind>(GeneratedCodeRecognizer generatedCod
     {
         if (HasMatchingScope(AnalysisContext.SupportedDiagnostics))
         {
+            ConcurrentDictionary<SyntaxTree, bool> canProceedAnalysisCache = new();
             Context.RegisterSyntaxNodeAction(x =>
-                    Execute<SonarSyntaxNodeReportingContext, SyntaxNodeAnalysisContext>(
-                        new(AnalysisContext, x), action, x.Node.SyntaxTree, generatedCodeRecognizer),
-                syntaxKinds);
-        }
-    }
-
-    /// <inheritdoc cref="SonarAnalysisContext.Execute" />
-    private void Execute<TSonarContext, TRoslynContext>(TSonarContext context, Action<TSonarContext> action, SyntaxTree sourceTree, GeneratedCodeRecognizer generatedCodeRecognizer = null)
-        where TSonarContext : SonarAnalysisContextBase<TRoslynContext>
-    {
-        if (ShouldAnalyzeTree(sourceTree, generatedCodeRecognizer)
-            && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(AnalysisContext.SupportedDiagnostics, sourceTree)
-            && AnalysisContext.ShouldAnalyzeRazorFile(sourceTree))
-        {
-            action(context);
+            {
+                if (canProceedAnalysisCache.GetOrAdd(x.Node.SyntaxTree, CanProceedWithAnalysis))
+                {
+                    action(new(AnalysisContext, x));
+                }
+
+                bool CanProceedWithAnalysis(SyntaxTree tree) =>
+                    ShouldAnalyzeTree(tree, generatedCodeRecognizer)
+                    && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(AnalysisContext.SupportedDiagnostics, tree)
+                    && AnalysisContext.ShouldAnalyzeRazorFile(tree);
+            }, syntaxKinds);
```

Comment:
This block creates a delegate (for CanProceedWithAnalysis) and a capture class (for `this`) on each invocation. We can avoid this on the happy path like so:
```suggestion
                if (!canProceedAnalysisCache.TryGetValue(x.Node.SyntaxTree, out var canProceedWithAnalysis))
                {
                    canProceedWithAnalysis = GetOrAddCanProceedWithAnalysis(generatedCodeRecognizer, canProceedAnalysisCache, x.Node.Tree);
                }
                if (canProceedWithAnalysis)
                {
                    action(new(AnalysisContext, x));
                }
            }, syntaxKinds);
            
            // Performance: Don't inline to avoid capture class and delegate allocations.
            bool GetOrAddCanProceedWithAnalysis(GeneratedCodeRecognizer generatedCodeRecognizer, ConcurrentDictionary<SyntaxTree, bool> canProceedAnalysisCache, SyntaxTree tree) =>
                canProceedAnalysisCache.GetOrAdd(tree, x => 
                    ShouldAnalyzeTree(tree, generatedCodeRecognizer)
                    && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(AnalysisContext.SupportedDiagnostics, tree)
                    && AnalysisContext.ShouldAnalyzeRazorFile(tree));

```